src/adapter/src/metrics.rs

```diff
@@ -142,6 +144,18 @@ impl Metrics {
                 name: "mz_webhook_get_appender_count",
                 help: "Count of getting a webhook appender from the Coordinator.",
             )),
+            check_scheduling_policies_seconds: registry.register(metric!(
+                name: "mz_check_scheduling_policies_seconds",
+                help: "The time each policy in `check_scheduling_policies` takes.",
+                var_labels: ["policy"],
+                buckets: histogram_seconds_buckets(0.000_128, 8.0),
+            )),
+            handle_scheduling_decisions_seconds: registry.register(metric!(
+                name: "mz_handle_scheduling_decisions_seconds",
+                help: "The time `handle_scheduling_decisions` takes.",
+                var_labels: ["altered_a_cluster"],
+                buckets: histogram_seconds_buckets(0.000_128, 8.0),
+            )),
```

Comment:
Ah geez, sorry I should have called this out earlier. Every [message execution is already instrumented with a tag of the message kind](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/coord.rs#L2594-L2601), so we don't need these.

Reply:
Hmm, that's nice! However, the custom metrics have two advantages over the generic ones:
- On the generic ones, the low end of the histogram is 0.128 seconds, while we'd like to measure much smaller delays here. (If I understood @benesch correctly.) The custom metrics' histogram goes down to 0.000128 seconds.
- The generic metric for `Message::SchedulingDecisions` includes both runs when `handle_scheduling_decisions` actually needs to turn a cluster on/off and when it doesn't need to touch a cluster. When it needs to touch a cluster, that is expected to take orders of magnitude more time than when it doesn't have to actually touch a cluster. We'd like to separately measure the case when no cluster is actually being touched, because that's the common case that will be happening every few seconds, so that is the case whose performance we were worried about.

Reply:
Done: I've added a label to `check_scheduling_policies_seconds` to indicate whether it's the main thread or the background task, and added an `observe` in the background task.

37790544b2aba4b1ddd0c45f16731e687bf8cd35

The following times are locally in debug mode with creating 10 clusters and 20 REFRESH MVs:
```
# HELP mz_check_scheduling_policies_seconds The time each policy in `check_scheduling_policies` takes.
# TYPE mz_check_scheduling_policies_seconds histogram
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.000128"} 0
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.000256"} 0
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.000512"} 0
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.001"} 16
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.002"} 127
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.004"} 165
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.008"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.016"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.032"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.064"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.128"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.256"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="0.512"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="1"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="2"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="4"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="8"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="background",le="+Inf"} 166
mz_check_scheduling_policies_seconds_sum{policy="refresh",thread="background"} 0.257805096
mz_check_scheduling_policies_seconds_count{policy="refresh",thread="background"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.000128"} 139
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.000256"} 154
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.000512"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.001"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.002"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.004"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.008"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.016"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.032"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.064"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.128"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.256"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="0.512"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="1"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="2"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="4"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="8"} 166
mz_check_scheduling_policies_seconds_bucket{policy="refresh",thread="main",le="+Inf"} 166
mz_check_scheduling_policies_seconds_sum{policy="refresh",thread="main"} 0.015206659999999993
mz_check_scheduling_policies_seconds_count{policy="refresh",thread="main"} 166
```
```
# HELP mz_handle_scheduling_decisions_seconds The time `handle_scheduling_decisions` takes.
# TYPE mz_handle_scheduling_decisions_seconds histogram
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.000128"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.000256"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.000512"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.001"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.002"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.004"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.008"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.016"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.032"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.064"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.128"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.256"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="0.512"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="1"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="2"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="4"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="8"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="false",le="+Inf"} 116
mz_handle_scheduling_decisions_seconds_sum{altered_a_cluster="false"} 0.003892188999999999
mz_handle_scheduling_decisions_seconds_count{altered_a_cluster="false"} 116
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.000128"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.000256"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.000512"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.001"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.002"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.004"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.008"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.016"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.032"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.064"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.128"} 0
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.256"} 13
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="0.512"} 19
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="1"} 32
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="2"} 50
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="4"} 51
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="8"} 51
mz_handle_scheduling_decisions_seconds_bucket{altered_a_cluster="true",le="+Inf"} 51
mz_handle_scheduling_decisions_seconds_sum{altered_a_cluster="true"} 41.313026674
mz_handle_scheduling_decisions_seconds_count{altered_a_cluster="true"} 51
```